# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/ChangeAttributeTransaction.cpp
# Commit: ce2087ddab5c
# Full Hash: ce2087ddab5c2efcbce3b52da02b3363a37d6813
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2020-04-04 09:29:13
# Regressor Bug: 1619914
# File Overlap Count: 1
# Description:
#   Bug 1619914 - part 1: Make each transaction class grab their members with local variable before touching the DOM tree r=m_kato
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D69302
# ==============================================================================

diff -r 1c8115a9a684 -r ce2087ddab5c editor/libeditor/ChangeAttributeTransaction.cpp
--- a/editor/libeditor/ChangeAttributeTransaction.cpp	Fri Apr 03 10:21:33 2020 +0000
+++ b/editor/libeditor/ChangeAttributeTransaction.cpp	Fri Apr 03 08:30:37 2020 +0000
@@ -63,36 +63,48 @@
 
   // Now set the attribute to the new value
   if (mRemoveAttribute) {
-    nsresult rv = mElement->UnsetAttr(kNameSpaceID_None, mAttribute, true);
+    OwningNonNull<Element> element = *mElement;
+    nsresult rv = element->UnsetAttr(kNameSpaceID_None, mAttribute, true);
     NS_WARNING_ASSERTION(NS_SUCCEEDED(rv), "Element::UnsetAttr() failed");
     return rv;
   }
 
-  nsresult rv = mElement->SetAttr(kNameSpaceID_None, mAttribute, mValue, true);
+  OwningNonNull<Element> element = *mElement;
+  nsresult rv = element->SetAttr(kNameSpaceID_None, mAttribute, mValue, true);
   NS_WARNING_ASSERTION(NS_SUCCEEDED(rv), "Element::SetAttr() failed");
   return rv;
 }
 
 NS_IMETHODIMP ChangeAttributeTransaction::UndoTransaction() {
+  if (NS_WARN_IF(!mElement)) {
+    return NS_ERROR_NOT_AVAILABLE;
+  }
   if (mAttributeWasSet) {
+    OwningNonNull<Element> element = *mElement;
     nsresult rv =
-        mElement->SetAttr(kNameSpaceID_None, mAttribute, mUndoValue, true);
+        element->SetAttr(kNameSpaceID_None, mAttribute, mUndoValue, true);
     NS_WARNING_ASSERTION(NS_SUCCEEDED(rv), "Element::SetAttr() failed");
     return rv;
   }
-  nsresult rv = mElement->UnsetAttr(kNameSpaceID_None, mAttribute, true);
+  OwningNonNull<Element> element = *mElement;
+  nsresult rv = element->UnsetAttr(kNameSpaceID_None, mAttribute, true);
   NS_WARNING_ASSERTION(NS_SUCCEEDED(rv), "Element::UnsetAttr() failed");
   return rv;
 }
 
 NS_IMETHODIMP ChangeAttributeTransaction::RedoTransaction() {
+  if (NS_WARN_IF(!mElement)) {
+    return NS_ERROR_NOT_AVAILABLE;
+  }
   if (mRemoveAttribute) {
-    nsresult rv = mElement->UnsetAttr(kNameSpaceID_None, mAttribute, true);
+    OwningNonNull<Element> element = *mElement;
+    nsresult rv = element->UnsetAttr(kNameSpaceID_None, mAttribute, true);
     NS_WARNING_ASSERTION(NS_SUCCEEDED(rv), "Element::UnsetAttr() failed");
     return rv;
   }
 
-  nsresult rv = mElement->SetAttr(kNameSpaceID_None, mAttribute, mValue, true);
+  OwningNonNull<Element> element = *mElement;
+  nsresult rv = element->SetAttr(kNameSpaceID_None, mAttribute, mValue, true);
   NS_WARNING_ASSERTION(NS_SUCCEEDED(rv), "Element::SetAttr() failed");
   return rv;
 }